File: 백준/Gold/14938. 서강그라운드/서강그라운드.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, m, r, max = Integer.MIN_VALUE, itemCount;
    static int[] weightSum, items;
    static ArrayList<ArrayList<WeightedEdge>> graph;

    public static void main(String[] args) throws IOException {
        setVariable();
        PriorityQueue<WeightedEdge> queue = new PriorityQueue<>(Comparator.comparing(a -> a.weight));
        for (int i = 0; i < n; i++) {
            queue.add(new WeightedEdge(i + 1, 0));
            initWeightSum();
            weightSum[i] = 0;
            while (!queue.isEmpty()) {
                WeightedEdge tmp = queue.poll();
                if (weightSum[tmp.nextVertex - 1] < tmp.weight) continue;
                for (WeightedEdge edge:graph.get(tmp.nextVertex - 1)) {
                    if (weightSum[edge.nextVertex - 1] > tmp.weight + edge.weight) {
                        weightSum[edge.nextVertex - 1] = tmp.weight + edge.weight;
                        queue.add(new WeightedEdge(edge.nextVertex, weightSum[edge.nextVertex - 1]));
                    }
                }
            }
            itemCount = 0;
            for (int j = 0; j < n; j++) {
                if (weightSum[j] <= m) {
                    itemCount += items[j];
                }
            }
            max = Math.max(max, itemCount);
        }
        System.out.println(max);
    }
    static void setVariable() throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());
        weightSum = new int[n];
        items = new int[n];
        st = new StringTokenizer(br.readLine());
        graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            items[i] = Integer.parseInt(st.nextToken());
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < r; i++) {
            st = new StringTokenizer(br.readLine());
            int v1 = Integer.parseInt(st.nextToken());
            int v2 = Integer.parseInt(st.nextToken());
            int wt = Integer.parseInt(st.nextToken());
            graph.get(v1 - 1).add(new WeightedEdge(v2, wt));
            graph.get(v2 - 1).add(new WeightedEdge(v1, wt));
        }
    }
    static void initWeightSum() {
        for (int i = 0; i < n; i++) {
            weightSum[i] = 3001;
        }
    }
}
class WeightedEdge {
    int nextVertex;
    int weight;
    public WeightedEdge(int nextVertex, int weight) {
        this.nextVertex = nextVertex;
        this.weight = weight;
    }
}

File: 백준/Gold/14938. 서강그라운드/서강그라운드.cc
#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;

typedef struct weighted_edge {
    int next_vtx, weight;
    weighted_edge(int next_vtx, int wt) : next_vtx(next_vtx), weight(wt) {}

    bool operator<(const weighted_edge e) const {
        return this->weight > e.weight;
    }
} WeightedEdge;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int n, m, r;
    cin >> n >> m >> r;

    vector<vector<WeightedEdge>> graph(n);

    int items[n];
    for (int i = 0; i < n; i++) {
        cin >> items[i];
    }

    int u, v, w;
    for (int i = 0; i < r; i++) { // 그래프를 저장하는 최선의 방법인가;
        cin >> u >> v >> w;

        graph[u - 1].push_back(WeightedEdge(v, w));
        graph[v - 1].push_back(WeightedEdge(u, w));
    }

    priority_queue<WeightedEdge> pque;
    int weight_sum[n];

    int answer = 0;
    for (int i = 0; i < n; i++) {
        pque.push(WeightedEdge(i + 1, 0));

        fill_n(weight_sum, n, 3001);
        weight_sum[i] = 0;

        while (!pque.empty()) {
            WeightedEdge tmp = pque.top(); pque.pop(); // poll()이 아니네?
            if (weight_sum[tmp.next_vtx - 1] < tmp.weight) continue;
            for (WeightedEdge edge : graph[tmp.next_vtx - 1]) {
                if (weight_sum[edge.next_vtx - 1] > tmp.weight + edge.weight) {
                    weight_sum[edge.next_vtx - 1] = tmp.weight + edge.weight;
                    pque.push(WeightedEdge(edge.next_vtx, weight_sum[edge.next_vtx - 1]));
                }
            }
        }

        int itemCount = 0;
        for (int j = 0; j < n; j++) {
            if (weight_sum[j] <= m) {
                itemCount += items[j];
            }
        }
        answer = max(answer, itemCount);
    }

    cout << answer << '\n';

    return 0;
}